libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/internal/ui/LocationPuckComponent.kt

```diff
@@ -21,8 +27,14 @@ class LocationPuckComponent(
             if (getLocationProvider() != locationProvider) {
                 setLocationProvider(locationProvider)
             }
-            locationPuck = this@LocationPuckComponent.locationPuck
+            locationPuck = contractProvider.get().getLocationPuck()
             enabled = true
         }
     }
 }
+
+internal class MapboxLocationPuckComponentContract(
+    private val locationPuckOptions: LocationPuckOptions
+) : LocationPuckComponentContract {
+    override fun getLocationPuck(): LocationPuck = locationPuckOptions.freeDrivePuck
```

Comment:
I don't really like that we use only `freeDrivePuck` in this case. I think `LocationPuckComponent` should accept `LocationPuck` like it was before.

Reply:
> I don't really like that we use only freeDrivePuck in this case.
To be able to reference different pucks from `LocationPuckOptions`, `MapboxLocationPuckComponentContract` needs to understand different `navigationState`. This info is not available outside Drop-In at the moment. All default `Mapbox*Contract` for components like `RouteLine`, `RouteArrow` etc suffer from the same problem. Their default implementations are mostly placeholders at the moment to get the functionality to work. 

> I think LocationPuckComponent should accept LocationPuck like it was before.

In this case how would we pass in puck for different navigation states? If we want to inject the puck in the component then the only other solution I can see is writing different `LocationPuckComponents` like `FreeDriveLocationPuckComponent`, `ActiveNavigationLocationPuckComponent` etc.

Reply:
> In this case how would we pass in puck for different navigation states? If we want to inject the puck in the component then the only other solution I can see is writing different `LocationPuckComponents` like `FreeDriveLocationPuckComponent`, `ActiveNavigationLocationPuckComponent` etc.

You detach the old component and attach a new one. `MapViewBinder` already does that by using `reloadOnChange`, therefore I don't understand why we use a lambda here. Also since navigation state is not available outside Drop-In UI, it doesn't make sense to use `LocationPuckOptions` outside Drop-In UI.

Reply:
With the location puck detaching the old component and reloading can happen in 2 scenarios:
- reload when navigation state changes
- reload when puck is changed in runtime
I already explored this option, but there are issues I am running into with this approach. If we keep the API(s) same as before:

```
ViewStyleCustomization

var locationPuck: LocationPuck? = null
```

```
MapViewBinder

override fun bind(viewGroup: ViewGroup): MapboxNavigationObserver {
    return navigationListOf(
        reloadOnChange(navigationState) { 
            LocationPuckComponent(location, provider) {
                LocationPuckComponentContractImpl(context)
            }
        }
    )
}
```

then in this case, we would have to assign location puck to each navigation state inside `LocationPuckComponentContractImpl`. The implementation would them become as follows:

```
LocationPuckComponentContractImpl

internal class LocationPuckComponentContractImpl(
    private val navigationViewContext: NavigationViewContext
) : LocationPuckComponentContract {
    override fun getLocationPuck(): LocationPuck {
        return when (navigationViewContext.store.state.value.navigation) {
            is NavigationState.FreeDrive -> {
                navigationViewContext.applyStyleCustomization {
                    locationPuck = LocationPuck2D(...)
                }
            ...
            ...
            }
        }
    }
}
```
In this case even if you are reloading the `LocationPuckComponent` every time the navigation state changes, you still need a way to support default location puck for each state and in cases when customer wants to inject it.

I also tried the following:

```
ViewStyleCustomization

var locationPuckFreeDrive: LocationPuck? = null
var locationPuckDestinationPreview: LocationPuck? = null
....
....
```
This would solve the issue above, but the `LocationPuckComponent` in that case would only reload when changing navigation state. Not when you are in the same state and changing location puck.

Reply:
@Zayankovsky I think I found a better solution. I have reverted the changes back to `LocationPuckComponent` still accepting a `locationPuck`, but still have `locationPuckOptions` coexist.

Reply:
That is actually what I had in mind, thanks.